[VM] Copy constant tensors to device

MasterJH5574 opened the pull request:
This PR improves the Relax VM to make it able to copy constants to backend devices (such as CUDA GPU). Prior to this PR, our VM only allocated on-device memory for intermediate tensors, while didn’t take constant tensors into account. And this PR solves the problem, with proper unit tests.

NOTE: Currently only single device is supported. In the future when we support multiple devices, we need to extend the constant copy as well.

cc @YuchenJin @ZihengJiang @yongwww

YuchenJin reviewed (commented):
Overall LGTM! Thanks @MasterJH5574!

Currently CI fails because we reuse `relay.Constant`, and in its constructor the `shape_` and `checked_type_`  is not filled, so it cannot be converted to a `te.Tensor`. There are three options:
1. Initialize `shape_` and `checked_type_` of `relay.Constant` in its constructor;
2. Have a `relax.Constant` to separate it from `relay.Constant`;
3. Check if the Expr is Constant in [TETensor](https://github.com/tlc-pack/relax/blob/relax/src/relax/ir/emit_te.cc#L39), if yes, get the shape and dtype from the Constant's data.

Before #116, we took Option 1, but we changed to fill the checked type and shape in Normalizer in that PR, without considering this emit_te case. Option 3 is not a bad starting point, we can always switch to other options if there are needs in the future.

psrivas2 commented on src/runtime/relax_vm/vm.cc:
```diff
@@ -47,6 +58,17 @@ PackedFunc VirtualMachine::GetFunction(const std::string& name,
       this->Init(devices, alloc_types);
+
+      // Copy NDArray constants to the devices
+      // TODO(tvm-team): support multiple devices
```
In case of multiple devices should we fail with a useful error message that multiple devices is not supported yet and preferably add a failing test too?

MasterJH5574 replied:
We do have error message here:

https://github.com/tlc-pack/relax/blob/40720e913c4fb81852aacfc2bf6dcd07f4c7228f/src/runtime/relax_vm/vm.cc#L92-L93

But I don’t think we’re supposed to have such failing tests, since the multi-device support will definitely be introduced one day. CC @YuchenJin what do you think?

YuchenJin replied:
I agree, it's a TODO item in the future. :)

psrivas2 commented on tests/python/relax/test_vm.py:
```diff
@@ -634,6 +634,55 @@ def expected_output():
     tvm.testing.assert_allclose(res.numpy(), expected_output(), rtol=1e-7, atol=1e-7)
 
 
+def test_vm_emit_te_constant_param_cpu():
```
This test fails for me when I checkout the PR. The error is same as CI failure error:

`ValueError: Check failed: (shape_expr->IsInstance<ShapeExprNode>()) is false: Expression does not have an known symbolic shape, please consider use match_shape to constrain the shape before passing into te_tensor`

MasterJH5574 replied:
Yes thanks. The problem is solved :-)

psrivas2 commented on tests/python/relax/test_vm.py:
```diff
@@ -634,6 +634,55 @@ def expected_output():
+
+
+@tvm.testing.requires_gpu
+def test_vm_emit_te_constant_param_gpu():
```
This test and the _cpu test above have a lot of code in common (building IR, asserts). Can we extract that out in a separate function. For example, https://github.com/tlc-pack/relax/blob/relax/tests/python/relax/test_autotir_integration.py#L219

psrivas2 reviewed (changes requested):
Thanks, looks good!

sunggg commented:
Thanks for your continuous great contributions, @MasterJH5574 
I have a few questions. 

- Can you give me a little more context how the device handled constant before this PR? If it didn't allocate on-device memory for constants, how did the execution access the constant values? 
- With this PR, which memory space on device are we copying constants to?
- Would this PR change the way to access the constants in the runtime (e.g., BYOC)?

MasterJH5574 commented:
> Currently CI fails because we reuse `relay.Constant`, and in its constructor the `shape_` and `checked_type_` is not filled, so it cannot be converted to a `te.Tensor`. There are three options:
> 
> 1. Initialize `shape_` and `checked_type_` of `relay.Constant` in its constructor;
> 2. Have a `relax.Constant` to separate it from `relay.Constant`;
> 3. Check if the Expr is Constant in [TETensor](https://github.com/tlc-pack/relax/blob/relax/src/relax/ir/emit_te.cc#L39), if yes, get the shape and dtype from the Constant's data.
> 
> Before #116, we took Option 1, but we changed to fill the checked type and shape in Normalizer in that PR, without considering this emit_te case. Option 3 is not a bad starting point, we can always switch to other options if there are needs in the future.

@YuchenJin Hey thanks for pointing it out! Sorry for not noticing the CI failure before.

As we’re moving to shape/checked-type inference in normalizer, the third option makes more sense than option 1. It’s now fixed in the latest comment.

But I’m still trying to understand the second option - what does separating the Constant node benefit?

MasterJH5574 commented:
> * Can you give me a little more context how the device handled constant before this PR? If it didn't allocate on-device memory for constants, how did the execution access the constant values?
> * With this PR, which memory space on device are we copying constants to?
> * Would this PR change the way to access the constants in the runtime (e.g., BYOC)?

@sunggg Thanks for the questions! 

* Q1. Prior to this PR, all constants reside in the host side (the CPU side). More specifically, prior to this PR we only have a constant pool in the Executable. https://github.com/tlc-pack/relax/blob/40720e913c4fb81852aacfc2bf6dcd07f4c7228f/include/tvm/runtime/relax_vm/executable.h#L129-L130

  In executables, we don’t have any information on the real devices where the execution runs, and thus we’re unable to copy the constant to the device side when building executables. We don’t have a constant pool in VM, and as a result, if the VM is executed on CPU then everything is okay. But once we want to execute a VM on CUDA GPU, the constants are not accessible to the VM and the execution fails.

* Q2. With this PR, we’re able to copy the constants to the device side when initializing the VM, since in the initialization we get a list of the devices. Say we have a CUDA GPU as device 0 and the host CPU as device 1. Then we’re copying the constants to the device 0 - the CUDA GPU.

  In the future we will support multiple devices. At that time different constants might be copied to different devices according to some memory plan. Let’s leave the work to the future.

* Q3. I don’t know much about your example about BYOC, and also, what do you mean by “the way to access the constants in the runtime”?

MasterJH5574 commented:
> We always use `DynTensorType` in Relax for the type of tensor, while Relay uses `TensorType`. If we take option 1, the checked_type_ of `relay.Constant` needs to be `DynTensorType` to be used in Relax, but it might affect Relay side.

@YuchenJin Thanks for the explanation! I suppose we’re only reusing some of the Relay Expr nodes and our building pipeline has no overlap with that on Relay side. So even if we use `DynTensorType` for the Relay Expr nodes in Relax, everything is supposed to be fine 🤔 ?

sunggg commented:
@MasterJH5574 Thanks for the info!
In Q2, I wondered if we can copy constants to the read-only memory in GPU. CUDA programming guide says the following:

> There are also two additional read-only memory spaces accessible by all threads: the constant and texture memory spaces. The global, constant, and texture memory spaces are optimized for different memory usages (see [Device Memory Accesses](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#device-memory-accesses)).

I just know these memories exist but do not know when/how to use them. It might worth ask folks who know about this. 


Please ignore Q3 - I didn't know current VM (without this PR) could not support GPU execution with constants.

MasterJH5574 commented on tests/python/relax/test_vm.py:
```diff
@@ -634,6 +634,55 @@ def expected_output():
+    mod = bb.get()
+    sch = tvm.tir.Schedule(mod, debug_mask="all")
+    loops = sch.get_loops(sch.get_block(name="T_add", func_name="add"))
+    sch.bind(loops[0], "threadIdx.x")
```
Good catch!  For CUDA backends, in this unit test, we bind a loop to `threadIdx.x` manually. In our E2E script, we chose to apply all the Meta-Schedule post processors to the IRModule that was not yet transformed (see [here](https://github.com/Hzfengsy/relax/blob/53d9f8bbe1fab337d52cc0c10376e72e43e57b13/tests/python/relax/e2e_autotir.py#L258)). You might think they are two different ways, but in fact, one of the post-processors for CUDA is to bind loops to threadIdx or blockIdx (the post-processor “Rewrite-Unbound-Block”) https://github.com/tlc-pack/relax/blob/40720e913c4fb81852aacfc2bf6dcd07f4c7228f/src/meta_schedule/postproc/rewrite_unbound_block.cc#L151-L163

The nature of both methods is to bind some loop to the GPU thread axes or block axes. This is because in Relax when we build an IRModule, we use `tvm.build(...)` to build the TIR PrimFuncs into a runtime library. https://github.com/tlc-pack/relax/blob/40720e913c4fb81852aacfc2bf6dcd07f4c7228f/python/tvm/relax/vm.py#L187

Due to this reason, if we don’t bind loops to threadIdx or blockIdx, the library build will fail :-( So I think at this moment, it’s necessary to have such a step of manual binding.

On Relay side, it doesn’t have the manual binding thing, but alternatively they apply a “lower-TE pass” on the IRModule being built. I observed that one behavior of this pass is to launch thread axes and bind loops to them, though the pass implementation looks complicated to me and I haven’t understood how they make it yet. https://github.com/tlc-pack/relax/blob/40720e913c4fb81852aacfc2bf6dcd07f4c7228f/src/relay/backend/graph_executor_codegen.cc#L233-L248

I do think in the future we should integrate the binding into build - let the building process, instead of users, automatically do this job. But for now, I prefer to do such a manual binding which is quick enough for us to unblock the E2E tuning on CUDA GPUs :-)

Let me know if you’re still confused of what I said :wink:

MasterJH5574 commented:
> In Q2, I wondered if we can copy constants to the read-only memory in GPU. CUDA programming guide says the following:
> 
> > There are also two additional read-only memory spaces accessible by all threads: the constant and texture memory spaces. The global, constant, and texture memory spaces are optimized for different memory usages (see [Device Memory Accesses](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#device-memory-accesses)).
> 
> I just know these memories exist but do not know when/how to use them. It might worth ask folks who know about this.

@sunggg Thanks for the question. Actually I don’t quite know how this is achieved as well LOL. I just dug into the `CopyTo` function here https://github.com/tlc-pack/relax/blob/fec50304329ba235e2505d91449886592fda3804/src/runtime/relax_vm/vm.cc#L30-L39
and found something interesting. Looks like we’re utilizing some CUDA APIs to implement a `CopyDataFromTo(...)` in `cuda_device_api.cc`. You can take a look at that part :-) https://github.com/tlc-pack/relax/blob/40720e913c4fb81852aacfc2bf6dcd07f4c7228f/src/runtime/cuda/cuda_device_api.cc#L140-L177

YuchenJin commented:
Thanks @MasterJH5574 for removing the last blocker for Relax e2e tuning on cuda! Thanks @psrivas2 @sunggg @yongwww for the reviews!
